lib/src/json/sink/byte_writer.dart

```diff
@@ -54,20 +55,21 @@ class JsonByteWriter implements JsonWriter<List<int>> {
   ///
   /// The resulting byte representation is a minimal JSON text with no
   /// whitespace between tokens.
-  JsonByteWriter(
-    Sink<List<int>> target, {
-    Encoding encoding = utf8,
-    bool? asciiOnly,
-  })  : _encoding = encoding,
+  JsonByteWriter(Sink<List<int>> target,
+      {Encoding encoding = utf8, bool? asciiOnly, bool allowReuse = false})
```

Comment:
> Looking at the actual code, I don't think it should work. The _sink is set to _Utf8EncoderSink(target) or _Utf8EncoderSink(_ByteAdapterSink(target)), depending on whether target implements ByteConversionSink already. Both of those classes forward a close to their wrapped sink, so the _sink.close() in line 69 below should end up calling _target.close(), whether we call _target.close() ourselves or not.)

The encoder is wrapped with `_NonClosingSink` and `_NonClosingSink` overrides `close` with empty body

Reply:
Wouldn't it be easier to wrap the `target` sink directly in a `_NonClosingSink` before creating the `JsonByteWriter`, instead of having the encoding do that only to the `_sink`? Then you don't need to change anything in the `JsonByteWriter` class, since the `_target.close()` would end up doing nothing.

If the `closeSink: false` argument doesn't actually *work* unless you use a special non-standard encoding, then I don't think the parameter belongs in the general `JsonByteWriter` API.

The alternative would be to add the non-closing sink wrapper _in_ this class, before storing `_target`, to ensure that it doesn't call `close` on the real target sink. (But then it really needs to check whether the original sink is a `ByteConversionSink`, and have a `_NonClosingByteConversionSink` too, to not throw away useful behavior that `Utf8Encoder.startChunkedConversion` uses).

It starts getting more complicated for what should have been a simple feature.

So, I'd recommend just creating a class:
```dart
class ReusableJsonByteWriter extends JsonByteWriter {
  ReusableJsonByteWriter(Sink<List<int>> target,
      {super.encoding = utf8, super.asciiOnly, super.allowReuse = false}): super(_NonClosingSink(target));
}
```
and use that instead of adding functionality that doesn't work by itself to the existing class.

Reply:
What does reusable mean here? Is the json writer reusable or the target sink can be used after json writer finishes. I meant the latter but feeling like it can mean both currently?

---

> If the closeSink: false argument doesn't actually work unless you use a special non-standard encoding, then I don't think the parameter belongs in the general JsonByteWriter API.

I did not understand this

> Wouldn't it be easier to wrap the target sink directly in a _NonClosingSink before creating the JsonByteWriter, instead of having the encoding do that only to the _sink? Then you don't need to change anything in the JsonByteWriter class, since the _target.close() would end up doing nothing.

I completely agree with this. This also suggest that other writers should also support this feature.

Reply:
The `ReusableJsonByteWriter` is not a class that I'd plan to add to this package, so you're free to call it what fits the purpose.

The point was that adding `allowReuse` or `closeSink` as proposed here won't actually *work* if you don't do something extra to make the `target` sink not do anything for `close` calls.
That makes it a bad API by itself, because it doesn't work. If I just write
```dart
 var writer = JsonByteWriter(someTarget, closeSink: false); // or allowReuse: true
```
then the target sink will still be closed by the call to `_sink.close()`.
And without the call to `_sink.close()`, and UTF-8 encoding might not be properly flushed, which is a different problem.

To make the API actually viable, the `JsonByteWriter` would have to do the `_NonClosingSink` wrapping itself.
That's not something I want to support on every class that accepts a sink. A non-closing sink is a concept that sink client shouldn't have to know about, they should just call `close` when they are done, because calling `close` is *how* they tell that they are done.

If you have another sink, and you want a `JsonByteWriter`, or any other sink client, to not close it when it's done, then wrapping the original sink in another, which forwards data to the original sink, and then notifies you when `close` is called, that's fine. That's perfectly good modular design, precisely because the sink client *doesn't* have to know about it.
So I'd be against adding extra non-closing sink support to any of the sink clients here, for them to even know about non-closing sinks is an abstraction failure. They should only need to know about `Sink` itself.

And then you, the user, can wrap your sinks any way you need to, at the abstraction level where that makes sense.

(Which, I guess, means that I don't think this PR should be accepted. It introduces a dependency on the concept of non-closing sinks where it doesn't belong. A sink client should always close its sink when it's done, and it's up to the sink provider to decide what that means.)

Reply:
> The point was that adding allowReuse or closeSink as proposed here won't actually work if you don't do something extra > to make the target sink not do anything for close calls.
>That makes it a bad API by itself, because it doesn't work. If I just write
> ```dart
>  var writer = JsonByteWriter(someTarget, closeSink: false); // or allowReuse: true
> ```
> then the target sink will still be closed by the call to _sink.close().
> And without the call to _sink.close(), and UTF-8 encoding might not be properly flushed, which is a different problem.

I don't think this is correct. First, in my local I already changed to continue working so it is working as expected. Second, if you check https://github.com/lrhn/json-tool/blob/master/lib/src/json/sink/byte_writer.dart#L64, the encoder is already wrapped with `_NonClosingSink` so it won't actually call `target.close` when `_sink.close` is called. That is why in https://github.com/lrhn/json-tool/blob/master/lib/src/json/sink/byte_writer.dart#L70 `_targe.close` is manually called.
Note that the codes from master.

> If you have another sink, and you want a JsonByteWriter, or any other sink client, to not close it when it's done, then > wrapping the original sink in another, which forwards data to the original sink, and then notifies you when close is called, > that's fine. That's perfectly good modular design, precisely because the sink client doesn't have to know about it. So I'd be against adding extra non-closing sink support to any of the sink clients here, for them to even know about non-closing sinks is an abstraction failure. They should only need to know about Sink itself.
>
> And then you, the user, can wrap your sinks any way you need to, at the abstraction level where that makes sense.

I believe most writers has a concept of providing a way to control how completion should happen, especially when they are low level apis. IMHO, this library is one of them and letting user to decide what should be the behavior of the writer is not a bad design decision.

What would you think about the following?

```dart
JsonWriter<String> jsonStringWriter(StringSink sink,
    {String? indent, bool asciiOnly = false, bool closeSink= true}) {
  if (!closeSink) {
    sink = NonClosingSink(sink);
  }
  if (indent == null) return JsonStringWriter(sink, asciiOnly: asciiOnly);
  return JsonPrettyStringWriter(sink, indent, asciiOnly: asciiOnly);
}

JsonWriter<List<int>> jsonByteWriter(Sink<List<int>> sink,
    {Encoding encoding = utf8, bool? asciiOnly, bool closeSink = true, bool closeSink = true}) {
  if (!closeSink) {
    sink = NonClosingSink(sink);
  }
  return JsonByteWriter(sink,
      encoding: encoding, asciiOnly: asciiOnly);
}
```

I am not so sure about `validateJsonSink` it already has `allowReuse` but I could understand how it is used